File: cses/04_graph_algorithms/1672_shortest-routes-ii/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần để tăng tốc.
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); m = int(data[pos + 1]); q = int(data[pos + 2])
    pos += 3

    # INF lớn hơn khoảng cách thật lớn nhất (~ (n-1)*1e9 < 2^40): mọi tổng đi qua một
    # cạnh không tồn tại vẫn >= INF, không tạo ra "đường tắt" giả.
    INF = 1 << 40

    # Ma trận khoảng cách: dist[i][i] = 0, còn lại khởi tạo INF.
    dist = [[INF] * n for _ in range(n)]
    for i in range(n):
        dist[i][i] = 0

    # Nạp các cạnh vô hướng; nếu có nhiều đường giữa cùng một cặp thì giữ đường nhỏ nhất.
    for _ in range(m):
        a = int(data[pos]) - 1
        b = int(data[pos + 1]) - 1
        c = int(data[pos + 2])
        pos += 3
        if c < dist[a][b]:
            dist[a][b] = c
            dist[b][a] = c

    # Floyd-Warshall: k là đỉnh trung gian. Vòng lặp trong cùng viết bằng list
    # comprehension + zip (chạy ở mức C) để nhanh nhất có thể trong pure Python.
    for k in range(n):
        dk = dist[k]
        for di in dist:
            dik = di[k]
            # Bỏ qua hàng khi không tới được k -> không thể cải thiện gì.
            if dik < INF:
                # di[j] = min(di[j], dik + dk[j]); walrus tính dik + dkj một lần.
                di[:] = [dij if dij < (s := dik + dkj) else s
                         for dij, dkj in zip(di, dk)]

    # Trả lời truy vấn trong O(1); còn INF nghĩa là không có đường -> in -1.
    out = []
    append = out.append
    for _ in range(q):
        a = int(data[pos]) - 1
        b = int(data[pos + 1]) - 1
        pos += 2
        d = dist[a][b]
        append(str(d) if d < INF else "-1")

    sys.stdout.write("\n".join(out))
    if out:
        sys.stdout.write("\n")


main()

File: cses/04_graph_algorithms/1672_shortest-routes-ii/solution.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m, q;
    std::cin >> n >> m >> q;

    // INF lớn hơn nhiều khoảng cách thật lớn nhất (~ (n-1)*1e9), chia 4 để cộng
    // hai giá trị INF cũng không tràn kiểu long long.
    constexpr long long INF = std::numeric_limits<long long>::max() / 4;

    // Ma trận khoảng cách: dist[i][i] = 0, còn lại khởi tạo INF.
    std::vector<std::vector<long long>> distance(
        n, std::vector<long long>(n, INF));
    for (int i = 0; i < n; ++i) {
        distance[i][i] = 0;
    }

    // Nạp các cạnh vô hướng; nếu có nhiều đường giữa cùng một cặp thì giữ đường nhỏ nhất.
    for (int i = 0; i < m; ++i) {
        int a, b;
        long long weight;
        std::cin >> a >> b >> weight;
        --a;
        --b;
        distance[a][b] = std::min(distance[a][b], weight);
        distance[b][a] = std::min(distance[b][a], weight);
    }

    // Floyd-Warshall: k là đỉnh trung gian được phép đi qua.
    for (int k = 0; k < n; ++k) {
        const std::vector<long long>& via = distance[k];
        for (int i = 0; i < n; ++i) {
            std::vector<long long>& row = distance[i];
            const long long to_k = row[k];
            // Bỏ qua hàng i nếu i không tới được k -> không thể cải thiện gì.
            if (to_k == INF) {
                continue;
            }
            // Cập nhật dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]).
            for (int j = 0; j < n; ++j) {
                const long long through_k = to_k + via[j];
                if (through_k < row[j]) {
                    row[j] = through_k;
                }
            }
        }
    }

    // Trả lời truy vấn trong O(1) bằng cách tra bảng; còn INF nghĩa là không có đường.
    for (int query = 0; query < q; ++query) {
        int a, b;
        std::cin >> a >> b;
        const long long answer = distance[a - 1][b - 1];
        std::cout << (answer == INF ? -1 : answer) << '\n';
    }
    return 0;
}
